bug: false 'builder escaped worktree' warning when working branch has staged changes

## Problem

The shepherd's builder validation emits a false-positive warning and fails when the working directory has staged/modified files — even when those changes are on a **non-main branch** (not the main branch that builders should work on).

## Observed behavior

During issue #2928's shepherd run, the validation failed with:

```
[WARN] Builder may have escaped worktree for issue #2928 — main branch has 1 uncommitted file(s)
[ERROR] builder phase validation failed: ... WARNING: main branch dirty (1 NEW files)
```

But `git status` showed:
```
On branch docs/guide-update-20260219
Changes to be committed:
  modified: WORK_LOG.md
  modified: WORK_PLAN.md
```

The repo was on `docs/guide-update-20260219` (a guide maintenance branch), not on `main`. The 'dirty' files were the guide's own staged changes — completely unrelated to the builder.

## Root cause

The validation check likely uses `git status` or checks the working tree for dirty files but doesn't verify that the current branch is actually `main`. It treats *any* worktree dirtiness as an escape indicator.

## Expected behavior

The 'builder escaped worktree' check should:
1. Only trigger when on the **main** branch (or the repo's default branch)
2. Ignore dirty state on other branches
3. Alternatively, scope the check to files that would be in the expected worktree path

## Impact

- Shepherd failed builder phase validation even though PR was successfully created
- Issue was left with `loom:blocked` label incorrectly
- Shepherd exited prematurely, requiring re-spawn

--- a/loom-tools/src/loom_tools/shepherd/phases/builder.py
+++ b/loom-tools/src/loom_tools/shepherd/phases/builder.py
@@ -3357,12 +3357,50 @@ class BuilderPhase:
         return paths.builder_log_file(ctx.config.issue)
 
     @staticmethod
+    def _get_repo_root_branch(ctx: ShepherdContext) -> str | None:
+        """Return the current branch name at repo_root, or None on failure."""
+        result = subprocess.run(
+            ["git", "-C", str(ctx.repo_root), "rev-parse", "--abbrev-ref", "HEAD"],
+            capture_output=True,
+            text=True,
+            check=False,
+        )
+        if result.returncode == 0 and result.stdout.strip():
+            return result.stdout.strip()
+        return None
+
+    @staticmethod
+    def _get_default_branch(ctx: ShepherdContext) -> str:
+        """Return the default branch name (e.g. 'main') for the repo."""
+        result = subprocess.run(
+            ["git", "-C", str(ctx.repo_root), "symbolic-ref", "refs/remotes/origin/HEAD"],
+            capture_output=True,
+            text=True,
+            check=False,
+        )
+        if result.returncode == 0 and result.stdout.strip():
+            return result.stdout.strip().replace("refs/remotes/origin/", "")
+        return "main"
+
+    @staticmethod
     def _snapshot_main_dirty(ctx: ShepherdContext) -> set[str]:
         """Snapshot main's dirty files (git status --porcelain lines).
 
         Called before the builder spawns so that _gather_diagnostics can
         distinguish pre-existing dirt from new files added by a worktree escape.
+
+        Returns an empty set when the repo root is not on the default branch —
+        dirty files on a non-main branch are not evidence of a builder worktree
+        escape and should never trigger the escape-detection logic.
         """
+        # Only snapshot when the repo root is actually on the default branch.
+        # If the shepherd is running on a feature/maintenance branch, staged
+        # changes on that branch must not be mistaken for a worktree escape.
+        current_branch = BuilderPhase._get_repo_root_branch(ctx)
+        default_branch = BuilderPhase._get_default_branch(ctx)
+        if current_branch is not None and current_branch != default_branch:
+            return set()
+
         result = subprocess.run(
             ["git", "-C", str(ctx.repo_root), "status", "--porcelain"],
             capture_output=True,
@@ -3471,7 +3509,17 @@ class BuilderPhase:
         return None
 
     def _get_new_main_dirty_files(self, ctx: ShepherdContext) -> list[str]:
-        """Return list of NEW dirty files on main since baseline snapshot."""
+        """Return list of NEW dirty files on main since baseline snapshot.
+
+        Returns an empty list when the repo root is not on the default branch.
+        Dirty files on non-main branches (e.g. guide maintenance branches) are
+        not evidence of a builder worktree escape — see issue #2971.
+        """
+        current_branch = self._get_repo_root_branch(ctx)
+        default_branch = self._get_default_branch(ctx)
+        if current_branch is not None and current_branch != default_branch:
+            return []
+
         main_status = subprocess.run(
             ["git", "-C", str(ctx.repo_root), "status", "--porcelain"],
             capture_output=True, text=True, check=False,
@@ -3865,17 +3913,29 @@ class BuilderPhase:
         # Check if the builder accidentally modified files on main instead
         # of in the worktree.  Only flag files that are NEW since the builder
         # started — pre-existing dirty files are not evidence of escape.
-        main_status = subprocess.run(
-            ["git", "-C", str(ctx.repo_root), "status", "--porcelain"],
-            capture_output=True,
-            text=True,
-            check=False,
+        #
+        # Skip entirely when the repo root is not on the default branch.
+        # Dirty files on a non-main branch (e.g. a guide maintenance branch)
+        # are completely unrelated to the builder and must not be flagged as
+        # a worktree escape.  See issue #2971.
+        _repo_root_branch = self._get_repo_root_branch(ctx)
+        _default_branch = self._get_default_branch(ctx)
+        _on_default_branch = (
+            _repo_root_branch is None or _repo_root_branch == _default_branch
         )
+
         main_dirty_files: list[str] = []
-        if main_status.returncode == 0 and main_status.stdout.strip():
-            main_dirty_files = [
-                line for line in main_status.stdout.splitlines() if line
-            ]
+        if _on_default_branch:
+            main_status = subprocess.run(
+                ["git", "-C", str(ctx.repo_root), "status", "--porcelain"],
+                capture_output=True,
+                text=True,
+                check=False,
+            )
+            if main_status.returncode == 0 and main_status.stdout.strip():
+                main_dirty_files = [
+                    line for line in main_status.stdout.splitlines() if line
+                ]
 
         # Filter out pre-existing dirty files using the baseline snapshot
         if self._main_dirty_baseline is not None:

--- a/loom-tools/tests/shepherd/test_phases.py
+++ b/loom-tools/tests/shepherd/test_phases.py
@@ -15609,14 +15609,20 @@ class TestBuilderGatherDiagnosticsMainBranch:
             result = subprocess.CompletedProcess(
                 args=cmd, returncode=0, stdout="", stderr=""
             )
-            if "rev-parse" in cmd_str:
-                result.stdout = "feature/issue-42\n"
+            if "symbolic-ref" in cmd_str:
+                # Default branch detection — always return main
+                result.stdout = "refs/remotes/origin/main\n"
+            elif "rev-parse" in cmd_str and "--abbrev-ref" in cmd_str:
+                # Return "main" for repo root branch check, "feature/issue-42" for worktree
+                if str(wt_dir) in cmd_str:
+                    result.stdout = "feature/issue-42\n"
+                else:
+                    result.stdout = "main\n"
             elif "log" in cmd_str and "main..HEAD" in cmd_str:
                 result.stdout = ""
             elif "status --porcelain" in cmd_str:
                 # Check if this is the main branch check (uses repo_root)
                 # or worktree check (uses wt_dir)
-                cwd = kwargs.get("cwd")
                 if "-C" in cmd and str(tmp_path) in cmd_str and str(wt_dir) not in cmd_str:
                     # Main branch is dirty
                     result.stdout = " M src/lib.rs\n M src/parser.rs\n"
@@ -15658,8 +15664,13 @@ class TestBuilderGatherDiagnosticsMainBranch:
             result = subprocess.CompletedProcess(
                 args=cmd, returncode=0, stdout="", stderr=""
             )
-            if "rev-parse" in cmd_str:
-                result.stdout = "feature/issue-42\n"
+            if "symbolic-ref" in cmd_str:
+                result.stdout = "refs/remotes/origin/main\n"
+            elif "rev-parse" in cmd_str and "--abbrev-ref" in cmd_str:
+                if str(wt_dir) in cmd_str:
+                    result.stdout = "feature/issue-42\n"
+                else:
+                    result.stdout = "main\n"
             elif "log" in cmd_str and "main..HEAD" in cmd_str:
                 result.stdout = ""
             elif "status --porcelain" in cmd_str:
@@ -15699,8 +15710,13 @@ class TestBuilderGatherDiagnosticsMainBranch:
             result = subprocess.CompletedProcess(
                 args=cmd, returncode=0, stdout="", stderr=""
             )
-            if "rev-parse" in cmd_str:
-                result.stdout = "feature/issue-42\n"
+            if "symbolic-ref" in cmd_str:
+                result.stdout = "refs/remotes/origin/main\n"
+            elif "rev-parse" in cmd_str and "--abbrev-ref" in cmd_str:
+                if str(wt_dir) in cmd_str:
+                    result.stdout = "feature/issue-42\n"
+                else:
+                    result.stdout = "main\n"
             elif "log" in cmd_str and "main..HEAD" in cmd_str:
                 result.stdout = ""
             elif "status --porcelain" in cmd_str:
@@ -15747,8 +15763,15 @@ class TestBuilderMainDirtyBaseline:
             result = subprocess.CompletedProcess(
                 args=cmd, returncode=0, stdout="", stderr=""
             )
-            if "rev-parse" in cmd_str:
-                result.stdout = "feature/issue-42\n"
+            if "symbolic-ref" in cmd_str:
+                # Default branch detection — always return main
+                result.stdout = "refs/remotes/origin/main\n"
+            elif "rev-parse" in cmd_str and "--abbrev-ref" in cmd_str:
+                # Return "main" for repo root branch check, "feature/issue-42" for worktree
+                if str(wt_dir) in cmd_str:
+                    result.stdout = "feature/issue-42\n"
+                else:
+                    result.stdout = "main\n"
             elif "log" in cmd_str and "main..HEAD" in cmd_str:
                 result.stdout = ""
             elif "status --porcelain" in cmd_str:
@@ -15795,8 +15818,15 @@ class TestBuilderMainDirtyBaseline:
             result = subprocess.CompletedProcess(
                 args=cmd, returncode=0, stdout="", stderr=""
             )
-            if "rev-parse" in cmd_str:
-                result.stdout = "feature/issue-42\n"
+            if "symbolic-ref" in cmd_str:
+                # Default branch detection — always return main
+                result.stdout = "refs/remotes/origin/main\n"
+            elif "rev-parse" in cmd_str and "--abbrev-ref" in cmd_str:
+                # Return "main" for repo root branch check, "feature/issue-42" for worktree
+                if str(wt_dir) in cmd_str:
+                    result.stdout = "feature/issue-42\n"
+                else:
+                    result.stdout = "main\n"
             elif "log" in cmd_str and "main..HEAD" in cmd_str:
                 result.stdout = ""
             elif "status --porcelain" in cmd_str:
@@ -15842,8 +15872,15 @@ class TestBuilderMainDirtyBaseline:
             result = subprocess.CompletedProcess(
                 args=cmd, returncode=0, stdout="", stderr=""
             )
-            if "rev-parse" in cmd_str:
-                result.stdout = "feature/issue-42\n"
+            if "symbolic-ref" in cmd_str:
+                # Default branch detection — always return main
+                result.stdout = "refs/remotes/origin/main\n"
+            elif "rev-parse" in cmd_str and "--abbrev-ref" in cmd_str:
+                # Return "main" for repo root branch check, "feature/issue-42" for worktree
+                if str(wt_dir) in cmd_str:
+                    result.stdout = "feature/issue-42\n"
+                else:
+                    result.stdout = "main\n"
             elif "log" in cmd_str and "main..HEAD" in cmd_str:
                 result.stdout = ""
             elif "status --porcelain" in cmd_str:
@@ -15899,6 +15936,181 @@ class TestBuilderMainDirtyBaseline:
             baseline = builder._snapshot_main_dirty(mock_context)
 
         assert baseline == set()
+
+
+class TestBuilderNonMainBranchDirty:
+    """Test that worktree-escape detection is suppressed on non-main branches.
+
+    When the shepherd runs on a branch other than the default (e.g. a guide
+    maintenance branch like docs/guide-update), any staged or modified files
+    on that branch must NOT be treated as a builder worktree escape.
+    See issue #2971.
+    """
+
+    def test_non_main_branch_dirty_not_flagged(
+        self, mock_context: MagicMock, tmp_path: Path
+    ) -> None:
+        """Dirty files on a non-main repo-root branch should not set main_branch_dirty."""
+        wt_dir = tmp_path / "worktree"
+        wt_dir.mkdir()
+        mock_context.worktree_path = wt_dir
+        mock_context.config = ShepherdConfig(issue=42)
+        mock_context.repo_root = tmp_path
+
+        log_dir = tmp_path / ".loom" / "logs"
+        log_dir.mkdir(parents=True)
+
+        builder = BuilderPhase()
+
+        def fake_run(cmd, **kwargs):
+            cmd_str = " ".join(str(c) for c in cmd)
+            result = subprocess.CompletedProcess(
+                args=cmd, returncode=0, stdout="", stderr=""
+            )
+            if "symbolic-ref" in cmd_str:
+                # Default branch is main
+                result.stdout = "refs/remotes/origin/main\n"
+            elif "rev-parse" in cmd_str and "--abbrev-ref" in cmd_str:
+                if str(wt_dir) in cmd_str:
+                    result.stdout = "feature/issue-42\n"
+                else:
+                    # Repo root is on a non-main branch (guide maintenance)
+                    result.stdout = "docs/guide-update-20260219\n"
+            elif "log" in cmd_str and "main..HEAD" in cmd_str:
+                result.stdout = ""
+            elif "status --porcelain" in cmd_str:
+                if "-C" in cmd and str(tmp_path) in cmd_str and str(wt_dir) not in cmd_str:
+                    # Non-main branch has staged changes (guide's own changes)
+                    result.stdout = " M WORK_LOG.md\n M WORK_PLAN.md\n"
+                else:
+                    result.stdout = ""
+            elif "ls-remote" in cmd_str:
+                result.stdout = ""
+            elif "gh" in cmd_str:
+                result.stdout = "loom:building"
+            return result
+
+        with patch(
+            "loom_tools.shepherd.phases.builder.subprocess.run", side_effect=fake_run
+        ):
+            diag = builder._gather_diagnostics(mock_context)
+
+        assert diag["main_branch_dirty"] is False
+        assert diag["main_dirty_file_count"] == 0
+        assert diag["main_dirty_files"] == []
+        assert "WARNING" not in diag["summary"]
+
+    def test_non_main_branch_snapshot_returns_empty(
+        self, mock_context: MagicMock, tmp_path: Path
+    ) -> None:
+        """_snapshot_main_dirty should return empty set when repo root is not on main."""
+        mock_context.repo_root = tmp_path
+
+        builder = BuilderPhase()
+
+        def fake_run(cmd, **kwargs):
+            cmd_str = " ".join(str(c) for c in cmd)
+            result = subprocess.CompletedProcess(
+                args=cmd, returncode=0, stdout="", stderr=""
+            )
+            if "symbolic-ref" in cmd_str:
+                result.stdout = "refs/remotes/origin/main\n"
+            elif "rev-parse" in cmd_str and "--abbrev-ref" in cmd_str:
+                # Repo root is on a non-main branch
+                result.stdout = "docs/guide-update-20260219\n"
+            elif "status --porcelain" in cmd_str:
+                # Would return dirty files, but should be suppressed
+                result.stdout = " M WORK_LOG.md\n M WORK_PLAN.md\n"
+            return result
+
+        with patch(
+            "loom_tools.shepherd.phases.builder.subprocess.run", side_effect=fake_run
+        ):
+            baseline = builder._snapshot_main_dirty(mock_context)
+
+        # Snapshot should be empty — non-main dirty files must not be captured
+        assert baseline == set()
+
+    def test_non_main_branch_get_new_dirty_returns_empty(
+        self, mock_context: MagicMock, tmp_path: Path
+    ) -> None:
+        """_get_new_main_dirty_files should return [] when repo root is not on main."""
+        mock_context.repo_root = tmp_path
+
+        builder = BuilderPhase()
+
+        def fake_run(cmd, **kwargs):
+            cmd_str = " ".join(str(c) for c in cmd)
+            result = subprocess.CompletedProcess(
+                args=cmd, returncode=0, stdout="", stderr=""
+            )
+            if "symbolic-ref" in cmd_str:
+                result.stdout = "refs/remotes/origin/main\n"
+            elif "rev-parse" in cmd_str and "--abbrev-ref" in cmd_str:
+                # Repo root is on a non-main branch
+                result.stdout = "docs/guide-update-20260219\n"
+            elif "status --porcelain" in cmd_str:
+                # Would return dirty files, but should be suppressed
+                result.stdout = " M WORK_LOG.md\n M WORK_PLAN.md\n"
+            return result
+
+        with patch(
+            "loom_tools.shepherd.phases.builder.subprocess.run", side_effect=fake_run
+        ):
+            new_dirty = builder._get_new_main_dirty_files(mock_context)
+
+        assert new_dirty == []
+
+    def test_main_branch_dirty_still_detected(
+        self, mock_context: MagicMock, tmp_path: Path
+    ) -> None:
+        """Dirty files on the main branch should still be detected as worktree escape."""
+        wt_dir = tmp_path / "worktree"
+        wt_dir.mkdir()
+        mock_context.worktree_path = wt_dir
+        mock_context.config = ShepherdConfig(issue=42)
+        mock_context.repo_root = tmp_path
+
+        log_dir = tmp_path / ".loom" / "logs"
+        log_dir.mkdir(parents=True)
+
+        builder = BuilderPhase()
+
+        def fake_run(cmd, **kwargs):
+            cmd_str = " ".join(str(c) for c in cmd)
+            result = subprocess.CompletedProcess(
+                args=cmd, returncode=0, stdout="", stderr=""
+            )
+            if "symbolic-ref" in cmd_str:
+                result.stdout = "refs/remotes/origin/main\n"
+            elif "rev-parse" in cmd_str and "--abbrev-ref" in cmd_str:
+                if str(wt_dir) in cmd_str:
+                    result.stdout = "feature/issue-42\n"
+                else:
+                    # Repo root IS on main
+                    result.stdout = "main\n"
+            elif "log" in cmd_str and "main..HEAD" in cmd_str:
+                result.stdout = ""
+            elif "status --porcelain" in cmd_str:
+                if "-C" in cmd and str(tmp_path) in cmd_str and str(wt_dir) not in cmd_str:
+                    # Main branch has dirty files (builder escaped worktree)
+                    result.stdout = " M src/lib.rs\n?? new_file.py\n"
+                else:
+                    result.stdout = ""
+            elif "ls-remote" in cmd_str:
+                result.stdout = ""
+            elif "gh" in cmd_str:
+                result.stdout = "loom:building"
+            return result
+
+        with patch(
+            "loom_tools.shepherd.phases.builder.subprocess.run", side_effect=fake_run
+        ):
+            diag = builder._gather_diagnostics(mock_context)
+
+        assert diag["main_branch_dirty"] is True
+        assert diag["main_dirty_file_count"] == 2
+        assert "WARNING: main branch dirty" in diag["summary"]
 
 
 class TestBuilderGatherDiagnosticsNoChangesMarker:

